patch-db/src/model_paths.rs

```diff
@@ -0,0 +1,483 @@
+                    (None, None) => return true,
+                    (None, _) => return false,
+                    (_, None) => return true,
+                }
+            }
+        }
+        match (self, other) {
+            (JsonGlob::Path(x), JsonGlob::Path(y)) => x.starts_with(y),
+            (
+                JsonGlob::Path(x),
+                JsonGlob::PathWithStar(PathWithStar {
+                    segments: path,
+                    count: _,
+                }),
+            ) => starts_with_(
+                &x.iter()
+                    .map(|x| JsonGlobSegment::Path(x.to_string()))
+                    .collect(),
+                path,
+            ),
+            (
+                JsonGlob::PathWithStar(PathWithStar {
+                    segments: path,
+                    count: _,
+                }),
+                JsonGlob::Path(y),
+            ) => starts_with_(
+                path,
+                &y.iter()
+                    .map(|x| JsonGlobSegment::Path(x.to_string()))
+                    .collect(),
+            ),
+            (
+                JsonGlob::PathWithStar(PathWithStar {
+                    segments: path,
+                    count: _,
+                }),
+                JsonGlob::PathWithStar(PathWithStar {
+                    segments: path_other,
+                    count: _,
+                }),
+            ) => starts_with_(path, path_other),
+        }
+    }
+    /// When we need to convert back into a usuable pointer string that is used for the paths of the
+    /// get and set of the db.
+    pub fn as_pointer(&self, binds: &[&str]) -> JsonPointer {
+        match self {
+            JsonGlob::Path(json_pointer) => json_pointer.clone(),
+            JsonGlob::PathWithStar(PathWithStar {
+                segments: path,
+                count: _,
+            }) => {
+                let mut json_pointer: JsonPointer = Default::default();
+                let mut binds = binds.iter();
+                for path in (*path).iter() {
+                    match path {
+                        JsonGlobSegment::Path(path) => json_pointer.push_end(&path),
+                        JsonGlobSegment::Star => {
+                            if let Some(path) = binds.next() {
+                                json_pointer.push_end(path)
+                            }
+                        }
+                    }
+                }
+                json_pointer
+            }
+        }
+    }
+
+    pub fn star_count(&self) -> usize {
+        match self {
+            JsonGlob::Path(_) => 0,
+            JsonGlob::PathWithStar(PathWithStar { count, .. }) => *count,
+        }
+    }
+}
+
+/// Used to determine the phantom type for bindings, () for no binds, String for one, (String, String) for two, ...
+/// Also, there are message to indicate that this is not valid
+pub trait AsPhantom<SB> {
+    fn as_phantom_binds(&self) -> PhantomData<SB>;
+}
```

Comment:
Rename this to "AssertUnsaturatedArgs<Argset>" or something to that effect. Newtype `PhantomData<SB>` to `UnsaturatedArgs<SB>` or something like that. Do not export newtype constructor from this module. Have the trait be the only way to produce proofs of this type. Name the singular row for this trait as "assert_unused_args" or something like that.